src/ray/common/asio/instrumented_io_context.cc

```diff
@@ -80,63 +55,65 @@ void instrumented_io_context::post(std::function<void()> handler,
+  // handler stats it->second from multiple threads without acquiring a table-level
+  // readers lock in the callback.
+  // NOTE: We capture the handler stats and global stats shared pointers by value,
+  // ensuring that the reference count will be incremented and that the underlying
+  // structs won't be deallocated until all such handlers finish executing. This allows
+  // the handlers to execute without segfaulting even if the thread that constructed the
+  // io_context object exits without joining the handler threads (which can happen in a
+  // hard exit).
+  boost::asio::io_context::post([handler = std::move(handler), stats = it->second,
+                                 global_stats = global_stats_, name, start]() {
+    int64_t start_execution = absl::GetCurrentTimeNanos();
+    // Execute actual handler.
+    handler();
+    int64_t end_execution = absl::GetCurrentTimeNanos();
+    // Update queue and execution time stats.
+    const auto queue_time_ns = start_execution - start;
+    const auto execution_time_ns = end_execution - start_execution;
+    // Update handler-specific stats.
+    {
+      absl::MutexLock lock(&(stats->mutex));
+      // Handler-specific execution stats.
+      stats->stats.cum_execution_time += execution_time_ns;
+      // Handler-specific current count.
+      stats->stats.curr_count--;
+    }
+    // Update global stats.
+    {
+      absl::MutexLock lock(&(global_stats->mutex));
+      // Global queue stats.
+      global_stats->stats.cum_queue_time += queue_time_ns;
+      if (global_stats->stats.min_queue_time > queue_time_ns) {
+        global_stats->stats.min_queue_time = queue_time_ns;
+      }
+      if (global_stats->stats.max_queue_time < queue_time_ns) {
+        global_stats->stats.max_queue_time = queue_time_ns;
+      }
+    }
+  });
 }
 
 /// A helper for creating a snapshot view of the global stats.
 /// This acquires a reader lock on the provided global stats, and creates a
 /// lockless copy of the stats.
-inline GlobalStats to_global_stats_view(const GuardedGlobalStats &stats) {
-  absl::MutexLock lock(&(stats.mutex));
-  return GlobalStats(stats.stats);
+inline GlobalStats to_global_stats_view(std::shared_ptr<GuardedGlobalStats> stats) {
+  absl::MutexLock lock(&(stats->mutex));
+  return GlobalStats(stats->stats);
 }
 
 GlobalStats instrumented_io_context::get_global_stats() const {
   return to_global_stats_view(global_stats_);
 }
 
 /// A helper for creating a snapshot view of the stats for a handler.
-/// This acquires a reader lock on the provided guarded handler stats, and creates a
+/// This acquires a lock on the provided guarded handler stats, and creates a
 /// lockless copy of the stats.
-inline HandlerStats to_handler_stats_view(const GuardedHandlerStats &stats) {
-  absl::MutexLock lock(&(stats.mutex));
-  return HandlerStats(stats.stats);
+inline HandlerStats to_handler_stats_view(std::shared_ptr<GuardedHandlerStats> stats) {
```

Comment:
nit: I think the inline here is not necessary. It's cpp file, not included in other files.

Reply:
It was my [initial impression](https://github.com/ray-project/ray/pull/14546#discussion_r589860536) that these explicit `inline` specifiers weren't necessary. Would it be best practice here to remove the explicit `inline` specifier and put these helper functions into an anonymous namespace (to guarantee that it won't be externally visible)?

Reply:
I opted for the anonymous namespace: https://github.com/ray-project/ray/pull/14719/commits/ea65e28c5a6d15c6c9a4422c8ad153205f831220

Reply:
Actually, I think I'm wrong. `inline` is still useful here, which gives the compiler the hint for optimization. No inline can still make things compiled successfully though.

'anonymous namespace' is useful for the function defined in cc only since it'll be invisible from outside. I remember you can use inline and anonymous together, they are for different purpose.

Reply:
I thought that the anonymous namespace pattern was useful within non-header files when you're wanting to (1) prevent link collisions with your helper functions and other unnamespaced functions outside of the module, (2) make all such helpers translation unit local such that external linkage is impossible. Even if (2) doesn't serve as an extra inlining compiler hint, it will hopefully serve as a hint to future developers that want to publicly expose those helpers that they may want to consider whether they should add an explicit `inline` specifier given that they are opening the function up to external linkage.

Reply:
And I don't think that an explicit `inline` specifier would do anything more to encourage the compiler to inline the function than defining the function within an anonymous namespace (guaranteed no external linkage) in a non-header source file, right? The `inline` specifier is just a linker directive and is only useful when the function definition might show up in multiple translation units (i.e. defined in the header).